### DISCUSSION TITLE
Plot entity state over time

### DISCUSSION THREAD

**Discussion by kopytjuk** [facontidavide/PlotJuggler#470] [Ideas]
I have a message with multiple entities with each holding a state (e.g. location in cartesian plane, e.g. `e[i].x` and  `e[i].y`). Each entity has an id, i.e. `e[i].id`.

My message looks like `time, [e_1, e_2,  ..., e_n]`. The array length can change over time. The order of the entities is not determenistic, new entities (with new `id` can spawn).

My wish is to plot the `x` coordinate over time with a different color for each unique `id`.

![Untitled.png](https://user-images.githubusercontent.com/12762439/123612638-05e00100-d803-11eb-83ca-d1db18de057e.png)

Can I visualize such data in PlotJuggler? I played around with the UI and could not find a setting or set up the plot. If it is not possible how would one implement such functionality in the repository? I would love to contribute!

**Comment by facontidavide** [comment]
I need to know more about your data source. ROS? CSV? custom?

I can hardly understand how to reproduce and fix this without a tangible example.

In general, anyway, the solution is to create your own message parser.

**Reply by kopytjuk** [reply]
You are right, I should be more explicit. I use ROS and my application's message definition is the following:

`Entity` message:
```
int64  id
geometry_msgs/Point position
```

`Entity` message, this one I want to visualize over time:
```
std_msgs/Header header
Entity[] entities
```

I want to visualize the position (e.g. `x`) of unique entities (by `id`) over time.

**Reply by kopytjuk** [reply]
Currently we can plot it, but the `x` attribute is always over the `i`th element from `Entity[] entities` list. I.e. I want to group by `id` and plot afterwards.

**Reply by facontidavide** [reply]
In this case, I am afraid that you need to create your custom parser: https://github.com/PlotJuggler/plotjuggler-ros-plugins/tree/development/plugins/ros1_parsers

**Reply by facontidavide** [reply]
Note that the implementation is trivial, but being this a specific ROS message and not a generic one, such change will never be included in mainline repo.

**Reply by facontidavide** [reply]
This might be a good example similar to what you need

https://github.com/PlotJuggler/plotjuggler-ros-plugins/blob/development/plugins/ros1_parsers/jointstates_msg.h
